actix-web-location/src/domain.rs

```diff
@@ -0,0 +1,187 @@
+    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
+    pub dma: Option<u16>,
+
+    /// The name of the provider that produced this recommendation.
+    pub provider: String,
+}
+
+macro_rules! location_field {
+    ($field: ident, $type: ty) => {
+        location_field!(
+            $field,
+            $type,
+            concat!(
+                "Get an owned copy of the ",
+                stringify!($field),
+                ", or the default if the field is None"
+            )
+        );
+    };
+
+    ($field: ident, $type: ty, $doc: expr) => {
+        #[doc = $doc]
+        pub fn $field(&self) -> $type {
+            self.$field.clone().unwrap_or_default()
+        }
+    };
+}
+
+impl Location {
+    /// Create a builder for a [`Location`] that can be assembled incrementally.
+    pub fn build() -> LocationBuilder {
+        LocationBuilder::default()
+    }
+
+    location_field!(country, String);
+    location_field!(region, String);
+    location_field!(city, String);
+    location_field!(dma, u16);
+}
+
+#[derive(Default)]
+pub struct LocationBuilder {
+    country: Option<String>,
+    region: Option<String>,
+    city: Option<String>,
+    dma: Option<u16>,
+    provider: Option<String>,
+}
+
+macro_rules! builder_field {
+    ($field: ident, $type: ty) => {
+        pub fn $field<O: Into<Option<$type>>>(mut self, $field: O) -> Self {
+            self.$field = $field.into();
+            self
+        }
+    };
+}
+
+impl LocationBuilder {
+    builder_field!(country, String);
+    builder_field!(region, String);
+    builder_field!(city, String);
+    builder_field!(dma, u16);
+    builder_field!(provider, String);
+
+    pub fn finish(self) -> Result<Location, ()> {
+        Ok(Location {
+            country: self.country,
+            region: self.region,
+            city: self.city,
+            dma: self.dma,
+            provider: self.provider.ok_or(())?,
+        })
+    }
+}
+
+#[cfg(feature = "maxmind")]
+impl<'a> From<City<'a>> for LocationBuilder {
+    fn from(val: City<'a>) -> Self {
+        Location::build()
+            .country(
+                val.country
+                    .and_then(|country| country.iso_code)
+                    .map(String::from),
+            )
+            .region(
+                val.subdivisions
+                    // Subdivisions are listed in least-specific order. In the US, this might mean that subdivisions is state and then county. We want only the first.
+                    .and_then(|subdivisions| {
+                        subdivisions
+                            .get(0)
+                            .and_then(|subdivision| subdivision.iso_code)
+                    })
+                    .map(ToString::to_string),
+            )
+            .city(
+                val.city
+                    .and_then(|city| city.names)
+                    .and_then(|names| names.get("en").map(|name| name.to_string()))
```

Comment:
while we use "en" pretty exclusively, this seems like it could be a problem for future localization efforts, which we are going to need. Might it make sense to kinda sorta cheat and include the preferred language as part of a tuple? (e.g.  `From<(City, "en")>`?)

Reply:
Although I agree that we should leave room for localization where needed, I don't think this one of those places. This isn't data we plan to show to users, it's data we plan to use in our internal metrics and send to our internal partners, which presumably also expect English content.

>  future localization efforts, which we are going to need

What features do you foresee needing localized city names for?